caffe2/contrib/tensorrt/tensorrt_tranformer.cc

```diff
@@ -0,0 +1,423 @@
+#include "caffe2/contrib/tensorrt/tensorrt_tranformer.h"
+#include "caffe2/contrib/tensorrt/trt_utils.h"
+#include "caffe2/core/logging.h"
+#include "caffe2/core/operator.h"
+#include "caffe2/onnx/onnx_exporter.h"
+#include <onnx2trt.hpp>
+#include <NvInfer.h>
+
+#include <google/protobuf/text_format.h>
+#include <iostream>
+#include <unordered_set>
+
+namespace caffe2 {
+
+namespace {
+
+// TODO(yinghai): Remove the awkward conversion between unordered_map and map
+std::unordered_map<std::string, TensorShape> InferShapes(
+    NetDef* init_net,
+    NetDef* pred_net,
+    const std::unordered_map<std::string, TensorShape>& input_shape_hints) {
+  CaffeMap<std::string, TensorShape> shape_hints_ordered;
+  for (const auto& kv : input_shape_hints) {
+    shape_hints_ordered.emplace(kv.first, kv.second);
+  }
+  std::vector<std::unique_ptr<NetDef>> nets;
+  nets.emplace_back(init_net);
+  nets.emplace_back(pred_net);
+  InferBlobShapesAndTypes(shape_hints_ordered, nets);
+  for (auto& net : nets) {
+    net.release();
+  }
+  std::unordered_map<std::string, TensorShape> shape_hints;
+  for (const auto& kv : shape_hints_ordered) {
+    shape_hints.emplace(kv.first, kv.second);
+  }
+
+  return shape_hints;
+}
+
+std::vector<std::string> FigureInputs(
+    const NetDef& pred_net,
+    int start,
+    int end,
+    const std::vector<OperatorDef>& new_ops,
+    const std::unordered_set<std::string>& weights,
+    std::unordered_set<std::string>* initialization_list) {
+  // TODO: cache this
+  std::unordered_set<std::string> boundary_inputs;
+  for (const auto& i : pred_net.external_input()) {
+    boundary_inputs.emplace(i);
+  }
+
+  //
+  for (const auto& op : new_ops) {
+    for (const auto& output: op.output()) {
+      boundary_inputs.emplace(output);
+    }
+  }
+
+  std::unordered_set<std::string> total_inputs;
+  std::vector<std::string> total_inputs_vec;
+  for(int i = start; i < end; ++i) {
+    const auto& op = pred_net.op(i);
+    for (const auto& input: op.input()) {
+      auto rt = total_inputs.emplace(input);
+      if (rt.second) {
+        if (weights.count(input)) {
+          // We add weights as inputs too
+          total_inputs_vec.emplace_back(input);
+          initialization_list->emplace(input);
+        } else if (boundary_inputs.count(input)) {
+          LOG(INFO) << "Adding boundary input: " << input;
+          total_inputs_vec.emplace_back(input);
+        }
+      }
+    }
+  }
+  return total_inputs_vec;
+}
+
+std::vector<std::string>
+FigureOutputs(const NetDef& pred_net, int start, int end) {
+  std::unordered_set<std::string> all_outputs;
+  std::vector<std::string> all_outputs_vec;
+  std::unordered_set<std::string> ext_outputs;
+  for (const auto& e: pred_net.external_output()) {
+    ext_outputs.emplace(e);
+  }
+  for (int i = start; i < end; ++i) {
+    const auto& op = pred_net.op(i);
+    for (const auto& output: op.output()) {
+      all_outputs.emplace(output);
+      all_outputs_vec.emplace_back(output);
```

Comment:
You can have many duplicated in `all_outputs_vec`, which makes it inefficient. You should add only if `all_outputs` didn't have this element. Even more efficient would be to not have a vector in the first place, just iterate `all_outputs` and carefully use `std::unordered_set::erase`

Reply:
Yes, you are right. I don't know why I did it this way. :( I'll changes.